### PR TITLE
WIP: create resources for clusters to be managed by hive

### PR DESCRIPTION
**Pull request by dhellmann**
Define resources for 2 clusters that can be managed by hive running in
the primary cluster created by the normal scripts. Provide a cleanup
script to remove them.

Require the user to run the scripts separately from the Makefile,
because we don't always want the additional resources created.

Depends on https://github.com/metal3-io/metal3-dev-env/pull/231

### CODE REVIEW

File: hive_setup.sh [hunk 1]
```diff
@@ -0,0 +1,278 @@
+#!/usr/bin/env bash
+set -xe
+
+source logging.sh
+source common.sh
+source utils.sh
+source rhcos.sh
+source ocp_install_env.sh
+
+function generate_hive_manifest() {
```
Comment (dhellmann) [inline hive_setup.sh]: @dgoodwin it would be nice to have your input on whether the data produced by this function looks right.

**Comment by dhellmann** [comment]
I'm still testing this locally, but it would be good to get some early review feedback.

File: hive_setup.sh [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+  provisioning:
+    installConfigSecretRef:
+      name: ${CLUSTER_NAME}-install-config
+    sshPrivateKeySecretRef:
+      name: ${CLUSTER_NAME}-ssh-private-key
+#    manifestsConfigMapRef:
+#      name: ${CLUSTER_NAME}-install-manifests
+    releaseImage: "${OPENSHIFT_RELEASE_IMAGE}"
+    sshKnownHosts:
+$(ssh-keyscan -H ${PROVISIONING_HOST_IP} 2>/dev/null | sed -e 's/^/      - "/g' -e 's/$/"/g')
```
Comment (dgoodwin) [inline hive_setup.sh]: I can't read this but if it's coming out as '      - "SSHKEY"' we're good. :)

File: hive_setup.sh [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    bareMetal:
+      libvirtSSHPrivateKeySecretRef:
+        name: ${CLUSTER_NAME}-ssh-private-key
+  provisioning:
+    installConfigSecretRef:
+      name: ${CLUSTER_NAME}-install-config
+    sshPrivateKeySecretRef:
+      name: ${CLUSTER_NAME}-ssh-private-key
+#    manifestsConfigMapRef:
+#      name: ${CLUSTER_NAME}-install-manifests
```
Comment (dgoodwin) [inline hive_setup.sh]: I assume you're targeting 4.4 thus no install manifests needed.

File: hive_setup.sh (hunk 2 above)
Comment (dhellmann) [reply hive_setup.sh]: On my host ssh-keyscan gives me

```
# localhost:22 SSH-2.0-OpenSSH_7.4
|1|LxI0OGI1Bcr97HslbGsrmWamUcM=|SnWlTu5a6+Fc+ZKtrw8VbKmqQjs= ssh-rsa AAAAB3NzaC1yc2EAAAADAQABAAABAQCh5D3FqClxnEZGZiZD9fclaVjkzZ98UX+xSBt+v4Q3ashpVhtYda61UTUmbXh9u8J44bYO5DS+OKtbDiBqQSB6p6VuxcIfjvapasUt7awAZIwBMUs2nYA9YA7F+0qpl14M+iK0zdbTqXhDn/KWfGxMK7YhZwPdmHNXV8V3B7yqmpc1skm3B4F5jgMWtY+a80e609SCtyVUaiRwibNILEvFM4YEMng9W9VAhweo1c4wt/hcwbLLQdu+XLrr+56WOy+9Ns9dfJBV4cHyV6Oflb1c2jcG/+FE+1jE1qpADP/lMaX9IPyil8sfPpBRWmpmy2La5pX8Wpc3zfMiz1VcYLAJ
# localhost:22 SSH-2.0-OpenSSH_7.4
|1|ZEpP9jvoFXWI9tbrTvrAlDXGlVE=|1CdDt0RJSixKsX6hRfbI4F1FVHE= ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBKWjJRzeUVuZs4yxSy4eu45xiANFIIbwE3e1aPzGD58x/NX7Yf+S8eFKq4RrsfSaK2hVJyJjvVIhUsU9z2sBJP8=
# localhost:22 SSH-2.0-OpenSSH_7.4
|1|rPeaRJpIzfn/i/ySG7ZHN3NdUjM=|xdcxdIXslQ/NQl7bIyJmToQ6gas= ssh-ed25519 AAAAC3NzaC1lZDI1NTE5AAAAIByDGt93ltGsuJWupi40PEn2VtiYuWw0/Bx2d9PYN1pQ
```

The comment lines go to stderr, so those are dropped.

The sed call adds some spaces and `- "` to the start of each line and `"` to the end of each line.

That gives us all of the host keys for the host in a YAML list.

File: hive_setup.sh (hunk 3 above)
Comment (dhellmann) [reply hive_setup.sh]: Yes, that's right. I haven't actually shown that this all works, yet, but I don't expect to need any extra manifests.

**Comment by stbenjam** [comment]
Ok so my initial thoughts about this PR: I would prefer something a little more flexible. I don't really like the hard coded number of clusters or how specific to Hive it is.

It also introduces duplicates of many of the values in common.sh. That affects the readability of that file a lot, it's already huge.  And I could see this being fragile too, as values don't get hive1 or 2 equivalents.

Could we just support doing multiple runs of dev-scripts to build up the infrastructure for multiple clusters? 

Just brainstorming, but on each run, a bunch of things would be randomly generated. For one, the assets dir would need to be different for each cluster (e.g. instead of always `ocp/`). Then we could store dev-scripts state in there for each particular cluster. Each dev-scripts "deployment" would generate distinct interfaces, networks, and VM's.  E.g. one run creates `ocp-xpqc/` directory, baremetal-xpqc and provision-xpqc interfaces, configured for 192.29.0.0/24 and 172.0.0.0/24, respectively. The next run of dev-scripts `make` creates something completely different.  `make clean` would need an argument to destroy a particular "deployment" (e.g. `DEPLOYMENT=xpqc make clean`).  If the part where we run the cluster install was optional, it'd let you create as many sets of "hardware" as you wanted and then you could use Hive to manage them.

There's a strong use case for this outside of Hive. I would love to be able to deploy multiple dev-scripts clusters on my dev box. IIRC, @hardys had some look at this a while back and had a PoC to do it, but now that nearly everything is configurable it's much more possible.

What do you think?

**Comment by dhellmann** [comment]
> Ok so my initial thoughts about this PR: I would prefer something a little more flexible. I don't really like the hard coded number of clusters or how specific to Hive it is.
> 
> It also introduces duplicates of many of the values in common.sh. That affects the readability of that file a lot, it's already huge. And I could see this being fragile too, as values don't get hive1 or 2 equivalents.

Yeah, I didn't really like that, either.

> 
> Could we just support doing multiple runs of dev-scripts to build up the infrastructure for multiple clusters?

I think the main thing we would need for that is the ability to have different working directories as output (I think we have that, but I'm not sure it always works, I know we've seen issues) and to be able to name the bridges to use for the networks. The network naming changes wouldn't be terribly hard.

> Just brainstorming, but on each run, a bunch of things would be randomly generated. For one, the assets dir would need to be different for each cluster (e.g. instead of always `ocp/`). Then we could store dev-scripts state in there for each particular cluster. Each dev-scripts "deployment" would generate distinct interfaces, networks, and VM's. E.g. one run creates `ocp-xpqc/` directory, `baremetal-xpqc` and `provision-xpqc` interfaces, configured for `192.29.0.0/24` and `172.0.0.0/24`, respectively. The next run of dev-scripts `make` creates something completely different. `make clean` would need an argument to destroy a particular "deployment" (e.g. `DEPLOYMENT=xpqc make clean`). If the part where we run the cluster install was optional, it'd let you create as many sets of "hardware" as you wanted and then you could use Hive to manage them.

What if we just add the cluster name to all of those things? Then the output is deterministic as well as distinct.

> There's a strong use case for this outside of Hive. I would love to be able to deploy multiple dev-scripts clusters on my dev box. IIRC, @hardys had some look at this a while back and had a PoC to do it, but now that nearly everything is configurable it's much more possible.
> 
> What do you think? It might be too grandiose a vision, but something I'd really like to see dev-scripts move towards.

The other thing I'm trying to do in this set of patches *is* to make it specific to Hive, though. There are a few assets to be generated to make a Hive deployment work. It's not extremely complicated, but it's non-trivial, so I was hoping to automate it.

**Comment by stbenjam** [comment]
> > Ok so my initial thoughts about this PR: I would prefer something a little more flexible. I don't really like the hard coded number of clusters or how specific to Hive it is.
> > It also introduces duplicates of many of the values in common.sh. That affects the readability of that file a lot, it's already huge. And I could see this being fragile too, as values don't get hive1 or 2 equivalents.
> 
> Yeah, I didn't really like that, either.
> 
> > Could we just support doing multiple runs of dev-scripts to build up the infrastructure for multiple clusters?
> 
> I think the main thing we would need for that is the ability to have different working directories as output (I think we have that, but I'm not sure it always works, I know we've seen issues) and to be able to name the bridges to use for the networks. The network naming changes wouldn't be terribly hard.
>
> > Just brainstorming, but on each run, a bunch of things would be randomly generated. For one, the assets dir would need to be different for each cluster (e.g. instead of always `ocp/`). Then we could store dev-scripts state in there for each particular cluster. Each dev-scripts "deployment" would generate distinct interfaces, networks, and VM's. E.g. one run creates `ocp-xpqc/` directory, `baremetal-xpqc` and `provision-xpqc` interfaces, configured for `192.29.0.0/24` and `172.0.0.0/24`, respectively. The next run of dev-scripts `make` creates something completely different. `make clean` would need an argument to destroy a particular "deployment" (e.g. `DEPLOYMENT=xpqc make clean`). If the part where we run the cluster install was optional, it'd let you create as many sets of "hardware" as you wanted and then you could use Hive to manage them.
> 
> What if we just add the cluster name to all of those things? Then the output is deterministic as well as distinct.

:+1: That would also make it easy to see what belongs to what.

> > There's a strong use case for this outside of Hive. I would love to be able to deploy multiple dev-scripts clusters on my dev box. IIRC, @hardys had some look at this a while back and had a PoC to do it, but now that nearly everything is configurable it's much more possible.
> > What do you think? It might be too grandiose a vision, but something I'd really like to see dev-scripts move towards.
> 
> The other thing I'm trying to do in this set of patches _is_ to make it specific to Hive, though. There are a few assets to be generated to make a Hive deployment work. It's not extremely complicated, but it's non-trivial, so I was hoping to automate it.

Ah right, I think having that part specific to Hive is fine, but do you think it's feasible to avoid needing the extra set of HIVE#_ vars?

**Pull request closed without merging**